asserts/asserts.go

```diff
@@ -0,0 +1,285 @@
+	// the authority that signed this assertion
+	AuthorityID() string
+
+	// Header retrieves the header with name
+	Header(name string) string
+
+	// the body of this assertion
+	Body() []byte
+
+	// Signature returns the signed content and its signature already split and decoded
+	Signature() (content []byte, sigtype string, signature []byte)
+}
+
+// AssertionBase is the concrete base to hold representation data for actual assertions.
+type AssertionBase struct {
+	headers map[string]string
+	body    []byte
+	// parsed revision
+	revision int
+	// preserved content
+	content []byte
+	// signature format/type
+	sigtype string
+	// decoded signature
+	sigpacket []byte
+}
+
+// Type returns the assertion type.
+func (ab *AssertionBase) Type() AssertionType {
+	return AssertionType(ab.headers["type"])
+}
+
+// Revision returns the assertion revision.
+func (ab *AssertionBase) Revision() int {
+	return ab.revision
+}
+
+// AuthorityID returns the authority-id a.k.a the signer id of the assertion.
+func (ab *AssertionBase) AuthorityID() string {
+	return ab.headers["authority-id"]
+}
+
+// Header returns the value of an header by name.
+func (ab *AssertionBase) Header(name string) string {
+	return ab.headers[name]
+}
+
+// Body returns the body of the assertion.
+func (ab *AssertionBase) Body() []byte {
+	return ab.body
+}
+
+// Signature returns the signed content and its signature already split and decoded.
+func (ab *AssertionBase) Signature() (content []byte, sigtype string, signature []byte) {
+	return ab.content, ab.sigtype, ab.sigpacket
+}
+
+// sanity check
+var _ Assertion = (*AssertionBase)(nil)
+
+var (
+	nlnl = []byte("\n\n")
+
+	// for basic sanity checking of header names
+	headerNameSanity = regexp.MustCompile("^[a-z][a-z0-9-]*[a-z0-9]$")
+)
+
+func parseHeaders(head []byte) (map[string]string, error) {
+	if !utf8.Valid(head) {
+		return nil, fmt.Errorf("header is not utf8")
+	}
+	headers := make(map[string]string)
+	for _, entry := range strings.Split(string(head), "\n") {
+		nameValueSplit := strings.Index(entry, ": ")
+		if nameValueSplit == -1 {
+			return nil, fmt.Errorf("header entry missing name value ': ' separation: %q", entry)
+		}
+		name := entry[:nameValueSplit]
+		if !headerNameSanity.MatchString(name) {
+			return nil, fmt.Errorf("invalid header name: %v", name)
+		}
+		headers[name] = entry[nameValueSplit+2:]
+	}
+	return headers, nil
+}
+
+// Decode parses a serialized assertion.
+//
+// The expected serialisation format looks like:
+//
+//   HEADER ("\n\n" BODY?)? "\n\n" SIGNATURE
+//
+// where:
+//
+//    HEADER is a set of header lines separated by "\n"
+//    BODY can be arbitrary,
+//    SIGNATURE is composed like:
+//      SIG-TYPE " " BASE64-SIG-PACKET
+//
+//    SIG-TYPE is the type/format expected for the base64
+//    encoded signature value, for now only "openpgp" is supported
+//
+// An header line looks like:
+//
+//   NAME ": " VALUE
+//
+// for sanity NAME is expected to match headerNameSanity regexp.
```

Comment:
That variable is not exported, so we shouldn't talk about it in the public API. It also feels like we probably don't need to document this. The field names will come out of specifications we write, so the internal validation is just making us honest about our own conventions.